### ISSUE TITLE
Crazy idea? Access to inverter registers from MQTT via grottproxy

### ISSUE THREAD

**Issue by eliasrm87** [johanmeijer/grott#402]
Hi,

I got my inverter installed by local installer company that utilizes the official Growatt platform for inverter monitoring and control. However, I'm interested in gaining control over the inverter while ensuring that the company remains under the impression that they have exclusive control. This necessity arises due to certain guarantee-related considerations.

While exploring grottserver's code, which enables reading and writing of registers through its web UI, an idea struck me. I'm curious whether it might be feasible to incorporate similar functionality into grottproxy. The objective would be to allow using MQTT to read and write registers, without sending any feedback to the official Growatt servers.

I'm eager to hear your insights. Could this idea be implemented? If so, I'd be grateful if you could provide guidance on the specific sections of code I should look at. Furthermore, I'm wondering whether there exists any documentation about to the communication protocol. Such information would be instrumental in comprehending the intricacies of the various commands at play. Should you possess such documentation, I would appreciative if you could share it, even if it is just a draft, which I could try to tidy up and share back.

Thank you for your dedication and hard work in maintaining this project.

**Comment by johanmeijer** [comment]
In a next version I want to integrtae grott en grottserver and then als create te possibility to sent commands in proxy mode. 

Using MQTT might be a good idea, but I do not think it will be much different form the HTTP API interface (you need a seperate thread to process the incomming commands). So from prograaming point of view it is a little bit the same. 

The basics of the protocol are described here:

https://www.vromans.org/johan/software/sw_growatt_wifi_protocol.html

This is only an older version. In the meanwhile to protocol has been changed  (is encrypted, has CRC checking there are new record types available ) but the basics are the same.

I do not have any other documentation available. Grottserver is the documentation for me (I am an old fashioned  IT guy  and still thinks code is obvious). Sorry for that. Maybe in the future when I have time I am going to document it..................

**Comment by eliasrm87** [comment]
Thank you very much, @johanmeijer!

I'm delighted to hear that you already have this in mind. I think you are correct; either way should work equally effectively, whether it's MQTT or a simple HTTP API request.

And thank you for sharing the protocol specification with me, even if it's not up to date. It should still be helpful as I dive into your code, and perhaps I can attempt to create an updated version while I learn about your implementation, if that's acceptable to you.

I'm interested in supporting this project, so please let me know if there are any "low-hanging fruit" tasks you think I could assist with while I become familiar with the code. While I'm not a Python expert, my expertise lies in C/C++, I believe I should be able to provide assistance.

On the other hand, I share a similar perspective with you regarding code clarity, but I've discovered the hard way that it's often true only for my own code. Therefore, I'm actively working on educating myself in methods to enhance my code documentation without overloading it with obvious comments.

**Comment by jmalm** [comment]
I did some investigations into how to code Time-of-Use setting messages. This is what I've came up with so far:

```python
import libscrc


def encode_tou(inverter_id, slot, use, enabled, start_hour, start_minute, end_hour, end_minute):
    """
    Returns instruction to set time-of-use.

    *slot*: The ToU slot to use (1 - 10)
    *use*: One of 'load', 'battery', and 'grid'
    *enabled*: True or False
    """
    data = bytes([1, 16]) # not sure what this means... Perhaps inverter ID?
    data += inverter_id.encode('ascii')
    # Pad with 0s to make it 32 bytes.
    data += bytes(32 - len(data))

    register1 = 220 + slot * 2 # I think these are 16-bit registers, where hour and minute are stored in the first and second bytes, respectively.
    register2 = 221 + slot * 2
    data += bytes([11, register1, 11, register2]) # not sure what the 11s mean...

    tou_data = [start_hour, start_minute, end_hour, end_minute]
    
    # Code *use* on bits 5 - 6 of start_hour.
    if use == 'load':
        # 'load' is 0.
        pass
    elif use == 'battery':
        # 'battery' is 1.
        tou_data[0] |= (1 << 5)
    elif use == 'grid':
        # 'grid' is 2.
        tou_data[0] |= (2 << 5)
    else:
        raise ValueError('Invalid use type. Should be one of "load", "battery", or "grid". Got "{}"'.format(use))
    
    # Code *enabled* on bit 7 of start_hour.
    if enabled:
        # 'enabled' is 1.
        tou_data[0] |= (1 << 7)

    data += bytes(tou_data)

    message = bytes([0, 1, 0, 6]) # not sure what this means... I think 06 is the protocol version.
    message += len(data).to_bytes(2, 'big') # length of data
    message += data
    message += libscrc.modbus(message).to_bytes(2, 'big')  # CRC -- NOTE: this is not the correct checksum.
    
    return message
```

For some reason I'm not getting the checksum (last two bytes) correct. Everything else appears to be correct when I compare with messages from the Shinephone app. Any ideas what could be wrong with the checksum?

For me, the Time-of-Use setting is the most important one for automation of when to charge / discharge the battery (based on solar forecast, load forecast, and current battery state-of-charge).

**Comment by typxxi** [comment]
> In a next version I want to integrtae grott en grottserver and then als create te possibility to sent commands in proxy mode.

Any updates on this topic ?

I need to update 8 inverters from time to with a new power ratio factor

**Comment by PedroKTFC** [comment]
> I did some investigations into how to code Time-of-Use setting messages. This is what I've came up with so far:
> 
> snip
>
> For me, the Time-of-Use setting is the most important one for automation of when to charge / discharge the battery (based on solar forecast, load forecast, and current battery state-of-charge).

Just came across this? Can you give an example so we can see what Shine is doing and what that gives? Thanks

**Comment by typxxi** [comment]
I guess he is talking about the registers for the times used to start and stop  charging here.
![image](https://github.com/johanmeijer/grott/assets/20662430/7f9b88cc-7b16-4f78-8258-9164f2f82192)

**Comment by anoppe** [comment]
> > In a next version I want to integrtae grott en grottserver and then als create te possibility to sent commands in proxy mode.
> 
> Any updates on this topic ?
> 
> I need to update 8 inverters from time to with a new power ratio factor

yeah, would be very nice feature to have. I've requested something alike here: https://github.com/johanmeijer/grott/discussions/529

**Comment by jmalm** [comment]
> I guess he is talking about the registers for the times used to start and stop  charging here.
> ![image](https://github.com/johanmeijer/grott/assets/20662430/7f9b88cc-7b16-4f78-8258-9164f2f82192)
> 

Yes, exactly. Those settings. I sent some from the mobile app and was able to intercept and reverse engineer them.

This was some time ago, and nowadays I'm using the battery for ancillary services (copilot translation...) in the Swedish electric grid.

**Comment by PedroKTFC** [comment]
I'm trying to dynamically change when and how my battery is charged/discharged. I'm using grottserver and set things using NodeRED in HA but I find the setting of registers, especially the start/stop time registers, inconsistent. Here's an example of one of my NodeRED flows that ensures my battery is sufficiently charged before the start of the peak price period. The Reg nnnn>m nodes set up register nnnn for the value m.
![image](https://github.com/johanmeijer/grott/assets/3314195/ce6d0155-1dde-42a4-964a-fe6925449ef7)

**Comment by jmalm** [comment]
I had another look at this. Looks like my settings are different from the screenshot by @typxxi . My time-slots are called Time-of-Use, and for each slot (10 in total), I can specify how to use the battery, start and end time, and whether the slot should be enabled.
![Screenshot_2024-11-02-14-38-28-558_com growatt shinephones-edit~2](https://github.com/user-attachments/assets/d9aed1d2-d2f7-4519-a056-88c7cbf18a27)

(Excuse the long screenshot...)

I've found that the registers for the Slot 1 are 3038 - 3039, Slot 2 3040 - 3041, and so on, ten pairs.
Enabled / disabled is the highest bit (bit 7) of the first byte of the first register position (16-bit positions).
Battery usage is in bits 5 - 6 of the first byte of the first register.
The start hour is in bits 0 - 4 of the first byte of the first register.
The start minute is in the second byte of the first register.
The end hour and minute is similarly in the first and second bytes of the second register, respectively.

**Comment by jmalm** [comment]
@johanmeijer, I'm not sure if the above is correct for all MOD TL3-XH inverters, but maybe it's something for your register document?

**Comment by jmalm** [comment]
I think my inverter firmware also has a setting for enabling / disabling VPP (Virtual Power Plant?) in register 30100, at least according to https://checkwatt.se/pdf/KP_installationsinstruktioner_GroWatt_MOD_3_30KTL3-XH.pdf. My inverter, which should be set to have VPP enabled, has a 1 in that register. I'm _guessing_ that that setting is somehow blocking my Time-of-Use settings to be used, but I haven't yet tested to turn it off.

**Comment by johanmeijer** [comment]
If you can set the parameters via growatt (server, phone app)  we can trace the communication flow via Grott and see which registers are set.

**Comment by jmalm** [comment]
@johanmeijer , yes, that's how I found the registers and format of the Time-of-Use settings. The VPP setting is only available via installer interface, I believe, which I don't have access to, so I can't verify that. That's also why I haven't yet tested it - I don't want to risk destroying something...

It would be nice to be able to GET  / PUT through grottserver, and _also_ be connected to server.growatt.com at the same time, to make / check settings in the app. It would make it easier to map out the settings for my inverter. Do you have any plans to support that development workflow?

I did notice that `grottserver.py` on the Beta-(2.8.x) branch is limited to registers up to 4096. The one in the `ledidobe/grottserver` docker image is even lower, 1125. Increasing this to beyond 30100, I was able to get the VPP value. What is the reason for the limit?
